New CLI arg "--param" to add parameters to the session.

sherfert opened the pull request:
cl:
Allows  providing Cypher parameters as command line arguments with `--param "name => value"`.

sherfert committed:
New CLI arg "--param" to add parameters to the session.

henriknyman commented on cypher-shell/src/main/java/org/neo4j/shell/CypherShell.java:
```diff
@@ -243,4 +208,12 @@ public String getActualDatabaseAsReportedByServer()
+    /**
+     * @return the parameter map.
+     */
+    public ParameterMap getParamaterMap()
```
Typo: `getParamaterMap` => `getParameterMap`

henriknyman commented on cypher-shell/src/main/java/org/neo4j/shell/commands/Param.java:
```diff
@@ -23,17 +23,26 @@
+     *                        If {@code false}, this param evaluates "--param" commands from CLI arguments.
+     */
+    public Param(@Nonnull final ParameterMap parameterMap,
+                 final boolean fromWithinShell) {
```
It feels a bit hacky to jack the command line parameter into the command with a boolean like this. It would be nicer to have a base class for the shared functionality of parsing and setting a parameter map (between 
ArgumentAction and Command), and leave the `Command` class as it is. I don't think `getName` and `getUsage` has any usages for command line argument actions.

henriknyman commented on cypher-shell/src/test/java/org/neo4j/shell/cli/CliArgHelperTest.java:
```diff
@@ -179,4 +187,43 @@ public void allowsEncryptionToBeTurnedOnOrOff() {
+
+    @Test
+    public void shouldParseSingleIntegerArgWithAddition() {
+        CliArgs cliArgs = CliArgHelper.parse( "-P", "foo=>3+5" );
```
It would be reassuring to have a test that also show that it works to have whitespaces, e.g. `foo => 3 + 5`

henriknyman reviewed (changes requested):
Looks pretty good in general, except for the way it abuses the param command class.

It also doesn't follow the code style and seem to have auto-formatted some existing code to a kind of half-way-neo4j code style.

sherfert committed:
Refactor Parameter setting code into its own class.

sherfert commented:
@henriknyman I have addressed your comments except the code style one. As long as we do not have some form of checkstyle, there is no point in discussing the style of the code if it is not enforced. I am not against having one either, but not passionate enough myself to introduce one.

henriknyman reviewed (commented):
@sherfert The new refactoring looks good! You just forgot to change the constructor call sites `new Param()` to remove the boolean, hence the failing tests.

About the code style, this project has always had a code style, the Java standard one, i.e. 
```
if (a == b) {
    doStuff(a);
}
```
rather than

```
if ( a == b )
{
    doStuff( a );
}
```

There could be a few anomalies where someone accidentally diverged from this by having auto-formatting with the Neo4j code style enabled.

I don't agree that you have to have an automated enforcement mechanism in place to follow a project's existing code style.

sherfert commented:
I fixed the compilation and I think all instances of where existing code was changed with just formatting have been removed.

sherfert committed:
Fix unrelated new test failures

because of removing deprecated features in Neo4j.
